### PR TITLE
Image color select

### PR DESCRIPTION
**Pull request by pejotejo**
## Introduced Changes

*Describe here what this pull request does*
-Created a twix-panel to get max/min rgb-values of multiple pixels
Fixes #

## ToDo / Known Issues
-changing the brush-size via scrolling overlaps with scrolling
*If this is a WIP describe which problems are to be fixed.*

## Ideas for Next Iterations (Not This PR)
*If there are some improvements that could be done in a next iteration, describe them here.*
-implement a Way to paint whole regions and get their values
-maybe histograms to get a colour distribution
- get the image in binary instead of .json
## How to Test

*Describe how to test your changes. (For the reviewer)*
-open twix and connect to a nao(preferably wired)

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -4,9 +4,9 @@ members = [
   "crates/aliveness",
   "crates/approx_derive",
   "crates/audio",
-  "crates/calibration",
   "crates/code_generation",
   "crates/communication",
+  "crates/calibration",
   "crates/constants",
```
Comment (h3ndrk) [inline Cargo.toml:10]: Please sort such lists alphabetically or is there a reason to reorder them?

**Review by h3ndrk — COMMENTED** [review]
This branch seems to include a lot of changes we already merged to `main`. Please rebase, thanks! 😊

File: tools/twix/src/main.rs [hunk 2]
```diff
@@ -145,7 +146,8 @@ impl_selectable_panel!(
     EnumPlotPanel,
     RemotePanel,
     TextPanel,
-    VisionTunerPanel
+    VisionTunerPanel,
+    ImageColorSelectPanel
```
Comment (schmidma) [inline tools/twix/src/main.rs:150]:
```suggestion
    ImageColorSelectPanel,
```
#hendrikComma

File: tools/twix/src/panels/image_color_select.rs [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+pub struct ImageColorSelectPanel {
+    nao: Arc<Nao>,
+    image_buffer: ValueBuffer,
+    cycler_selector: VisionCyclerSelector,
+    brush_size: f32,
+}
+struct PixelColor {
+    r: f32,
+    g: f32,
+    b: f32,
```
Comment (schmidma) [inline tools/twix/src/panels/image_color_select.rs:49]: please don't abbreviate

File: tools/twix/src/panels/image_color_select.rs [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+                let rgb = Rgb::from(ycbcr444);
+                [rgb.r, rgb.g, rgb.b, 255]
+            })
+            .collect::<Vec<_>>();
+        let image = ColorImage::from_rgba_unmultiplied([640, 480], &buffer);
+        Ok(image)
+    }
+}
+
+fn get_pixel_color(image: &ColorImage, pixel_pos: Point2<Pixel>) -> PixelColor {
```
Comment (schmidma) [inline tools/twix/src/panels/image_color_select.rs:264]: What are you computing here? The chromaticity?

**Review by schmidma — CHANGES_REQUESTED** [review]
minor things, didn't test that yet ;)

File: tools/twix/src/main.rs (hunk 2 above)
Comment (pejotejo) [reply tools/twix/src/main.rs]: done (needed to fix the panel makro with ole)

File: tools/twix/src/panels/image_color_select.rs (hunk 3 above)
Comment (pejotejo) [reply tools/twix/src/panels/image_color_select.rs]: renamed variables to red, green, blue

File: tools/twix/src/panels/image_color_select.rs [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+    let color32 = image.pixels[(pixel_pos.y() as usize) * 640 + (pixel_pos.x() as usize)];
+    let sum = color32.r() as f32 + color32.g() as f32 + color32.b() as f32;
+    PixelColor {
+        r: (color32.r() as f32) / sum,
+        g: (color32.g() as f32) / sum,
+        b: (color32.b() as f32) / sum,
```
[the remark this answers is not in the document]
Comment (pejotejo) [reply tools/twix/src/panels/image_color_select.rs]: if it's zero now, red, green and blue are set to zero too

File: tools/twix/src/panels/image_color_select.rs [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+                        pixel_pos.x() as usize,
+                        pixel_pos.y() as usize,
+                        cnt,
+                    ));
+                    ui.label(RichText::new("max:\t\t").strong());
+                    ui.colored_label(Color32::RED, format!("r: {:.3}", max.r));
+                    ui.colored_label(Color32::GREEN, format!("g: {:.3}", max.g));
+                    ui.colored_label(
+                        Color32::from_rgb(50, 150, 255),
+                        format!("b: {:.3}\n", max.b),
```
[the remark this answers is not in the document]
Comment (pejotejo) [reply tools/twix/src/panels/image_color_select.rs]: rewrote labels in a grid

File: tools/twix/src/panels/image_color_select.rs [hunk 7]
```diff
@@ -0,0 +1,272 @@
...
+                    r: 1.0,
+                    g: 1.0,
+                    b: 1.0,
+                };
+                let mut average = PixelColor {
+                    r: 0.0,
+                    g: 0.0,
+                    b: 0.0,
+                };
+                let mut cnt: usize = 0;
```
[the remark this answers is not in the document]
Comment (pejotejo) [reply tools/twix/src/panels/image_color_select.rs]: renamed variable to pixel_count

File: tools/twix/src/panels/image_color_select.rs [hunk 8]
```diff
@@ -0,0 +1,272 @@
...
+
+        let painter = TwixPainter::<Pixel>::paint_at(ui, response.rect).with_camera(
+            vector![640.0, 480.0],
+            Similarity2::identity(),
+            CoordinateSystem::LeftHand,
+        );
+
+        if let Some(hoverpos) = response.hover_pos() {
+            let pixel_pos = painter.transform_pixel_to_world(hoverpos);
+            if pixel_pos.x() < 640.0 && pixel_pos.y() < 480.0 {
```
[the remark this answers is not in the document]
Comment (pejotejo) [reply tools/twix/src/panels/image_color_select.rs]: done

**Comment by pejotejo** [comment]
Fixed doulbe transform_pixel_to_world. Now it should display the correct color values.

**Comment by schmidma** [comment]
@pejotejo What is the state on this feature? Should we do another iteration on the review? Or are you still working on this?

**Comment by pejotejo** [comment]
The panel itself is finished and I am already an a next iteration of it. So another round on the review would be nice.

**Comment by schmidma** [comment]
Sorry, for taking so long for the review... I'll do that now. Can you please rebase this PR to the newest main? Thanks :+1:

File: tools/twix/src/panels/image_color_select.rs [hunk 9]
```diff
@@ -0,0 +1,282 @@
...
+    YCbCr422,
+}
+
+pub struct ImageColorSelectPanel {
+    nao: Arc<Nao>,
+    image_buffer: ValueBuffer,
+    cycler_selector: VisionCyclerSelector,
+    brush_size: f32,
+}
+struct PixelColor {
```
Comment (schmidma) [inline tools/twix/src/panels/image_color_select.rs:46]:
```suggestion
}

struct PixelColor {
```

File: tools/twix/src/panels/image_color_select.rs [hunk 10]
```diff
@@ -0,0 +1,282 @@
...
+    red: f32,
+    green: f32,
+    blue: f32,
+}
```
Comment (schmidma) [inline tools/twix/src/panels/image_color_select.rs:50]: We already have a `color::Rgb` type, you can use that type for your colors

File: tools/twix/src/panels/image_color_select.rs [hunk 11]
```diff
@@ -0,0 +1,282 @@
...
+
+        let handle = ui
+            .ctx()
+            .load_texture("image", image.clone(), TextureOptions::default())
+            .id();
+        let texture = SizedTexture {
+            id: handle,
+            size: Vec2::new(image.width() as f32, image.height() as f32),
+        };
+        let response = ui.image(texture);
```
Comment (schmidma) [inline tools/twix/src/panels/image_color_select.rs:130]: This seems to result in the image begin of constant size. It is not scaling with the UI. I expect the image to fill the entire available space, similar to the `Image` panel

File: tools/twix/src/panels/image_color_select.rs [hunk 12]
```diff
@@ -0,0 +1,282 @@
...
+
+        ui.separator();
+
+        let painter = TwixPainter::<Pixel>::paint_at(ui, response.rect).with_camera(
+            vector![image.width() as f32, image.height() as f32],
+            Similarity2::identity(),
+            CoordinateSystem::LeftHand,
+        );
+
+        if let Some(hoverpos) = response.hover_pos() {
```
Comment (schmidma) [inline tools/twix/src/panels/image_color_select.rs:140]: please do not abbreviate

File: tools/twix/src/panels/image_color_select.rs [hunk 13]
```diff
@@ -0,0 +1,282 @@
...
+            let pixel_pos = painter.transform_pixel_to_world(hoverpos);
+            if pixel_pos.x() < image.width() as f32 && pixel_pos.y() < image.height() as f32 {
+                let scroll_delta = ui.input(|input| input.scroll_delta);
+                self.brush_size = (self.brush_size + scroll_delta[1]).clamp(1.0, 200.0);
+
+                let mut max = PixelColor {
+                    red: 0.0,
+                    green: 0.0,
+                    blue: 0.0,
+                };
```
Comment (schmidma) [inline tools/twix/src/panels/image_color_select.rs:150]: can you use helper functions or constructors, or constants like `PixelColor::BLACK`?

File: tools/twix/src/panels/image_color_select.rs [hunk 14]
```diff
@@ -0,0 +1,282 @@
...
+                for i in (pixel_pos.x() as isize - self.brush_size as isize)
+                    ..(pixel_pos.x() as isize + self.brush_size as isize + 1)
+                {
+                    for j in (pixel_pos.y() as isize - self.brush_size as isize)
+                        ..(pixel_pos.y() as isize + self.brush_size as isize + 1)
+                    {
+                        if f32::sqrt(
+                            f32::powi(i as f32 - pixel_pos.x(), 2)
+                                + f32::powi(j as f32 - pixel_pos.y(), 2),
+                        ) <= self.brush_size
```
Comment (schmidma) [inline tools/twix/src/panels/image_color_select.rs:172]: You can and should use iterators and `.filter` or similar functions to achieve this behavior ;) Look at https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.filter

File: tools/twix/src/panels/image_color_select.rs [hunk 15]
```diff
@@ -0,0 +1,282 @@
...
+                            min.green = min.green.min(color.green);
+                            min.blue = min.blue.min(color.blue);
+                            average.red += color.red;
+                            average.green += color.green;
+                            average.blue += color.blue;
+                            pixel_count += 1;
+                        }
+                    }
+                }
+                average.red /= pixel_count as f32;
```
Comment (schmidma) [inline tools/twix/src/panels/image_color_select.rs:192]: Is this dividing by zero?

File: tools/twix/src/panels/image_color_select.rs [hunk 16]
```diff
@@ -0,0 +1,282 @@
...
+        green: 0.0,
+        blue: 0.0,
+    };
+    if sum != 0.0 {
+        pixel.red = (color32.r() as f32) / sum;
+        pixel.green = (color32.g() as f32) / sum;
+        pixel.blue = (color32.b() as f32) / sum;
+    }
+    pixel
+}
```
Comment (schmidma) [inline tools/twix/src/panels/image_color_select.rs:282]: The `Rgb` struct already has such an implementation.

**Review by schmidma — CHANGES_REQUESTED** [review]

File: tools/twix/src/panels/image_color_select.rs (hunk 9 above)
Comment (pejotejo) [reply tools/twix/src/panels/image_color_select.rs]: done

File: tools/twix/src/panels/image_color_select.rs (hunk 15 above)
Comment (pejotejo) [reply tools/twix/src/panels/image_color_select.rs:193]: fixed

File: tools/twix/src/panels/image_color_select.rs (hunk 16 above)
Comment (pejotejo) [reply tools/twix/src/panels/image_color_select.rs:283]: same as in the previous, this is RGB and no rgb

**Review by schmidma — APPROVED** [review]

**Pull request merged**